proto/nodeinfo.go

```diff
@@ -17,34 +17,139 @@
 package proto
 
 import (
+	"errors"
 	"time"
 
+	ec "github.com/btcsuite/btcd/btcec"
+	pb "github.com/golang/protobuf/proto"
 	log "github.com/sirupsen/logrus"
 	"github.com/thunderdb/ThunderDB/crypto/asymmetric"
+	"github.com/thunderdb/ThunderDB/crypto/hash"
+	mine "github.com/thunderdb/ThunderDB/pow/cpuminer"
+	"github.com/thunderdb/ThunderDB/types"
 )
 
 var (
-	// NewNodeIDDifficulty is simply exposed for easy testing
+	// NewNodeIDDifficulty is exposed for easy testing
 	NewNodeIDDifficulty = 40
-	// NewNodeIDDifficultyTimeout is simply exposed for easy testing
+	// NewNodeIDDifficultyTimeout is exposed for easy testing
 	NewNodeIDDifficultyTimeout = 60 * time.Second
 )
 
+// NodeID is node name, will be generated from Hash(nodePublicKey)
+type NodeID string
+
+// AccountAddress is wallet address, will be generated from Hash(nodePublicKey)
+type AccountAddress string
+
+// NodeKey is node key on consistent hash ring, generate from Hash(NodeID)
+type NodeKey uint64
+
+// Node is all node info struct
+type Node struct {
+	ID        NodeID
+	Addr      string
+	PublicKey *ec.PublicKey
+	Nonce     mine.Uint256
+}
+
 // NewNode just return a new node struct
 func NewNode() *Node {
 	return &Node{}
 }
 
-// InitNodeCryptoInfo generate Node asymmetric key pair and generate Node.Nonce
-// Node.ID = Node.Nonce.Hash
+// NodeBytes is the marshal output of proto.Node.Marshal
+type NodeBytes []byte
+
+// Marshal the proto.Node with protobuf
+func (n *Node) Marshal() (nodeBytes NodeBytes, err error) {
+	if n == nil {
+		return nil, errors.New("nil node")
+	}
+	if n.PublicKey == nil {
+		return nil, errors.New("nil public key")
+	}
+
+	nodeBuf, err := pb.Marshal(&types.Node{
+		ID: &types.NodeID{
+			NodeID: string(n.ID),
+		},
+		Addr: n.Addr,
+		PublicKey: &types.PublicKey{
+			PublicKey: n.PublicKey.SerializeCompressed(),
+		},
+		Nonce: &types.Nonce{
+			A: n.Nonce.A,
+			B: n.Nonce.B,
+			C: n.Nonce.C,
+			D: n.Nonce.D,
+		},
+	})
+	if err != nil {
+		log.Errorf("marshal node failed: %s", err)
+		return nil, err
+	}
+
+	return NodeBytes(nodeBuf), err
+}
+
+// Unmarshal the proto.Node with protobuf
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported function UnmarshalNode should be of the form "UnmarshalNode ..."